File: Sorting/InsertionSort.py
#Worst-case performance	О(n2) comparisons and swaps
#Best-case performance	O(n) comparisons, O(1) swaps
#Average performance	О(n2) comparisons and swaps
#Worst-case space complexity	О(n) total, O(1) auxiliary


def insertionSort(arr): 
    for i in range(1, len(arr)): 
        k = arr[i] #k is to be inserted at proper place
        j = i-1
        while j >=0 and k < arr[j] : 
                arr[j+1] = arr[j] 
                j -= 1
        arr[j+1] = k 
    return arr

arr = [12, 11, 13, 5, 6] 
arr=insertionSort(arr) 
print ("Sorted array is: ", arr) 

File: Sorting/InsertionSort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

/*
Time Complexity
Worst Case- O(n*n)
Best Case- O(n) – When the array is already sorted
Space Complexity
Space Complexity of insertion sort is O(1)


 */

func main() {

	fmt.Println("Insertion Sort Demo ")
	rand.Seed(time.Now().UnixNano())
	min := 0
	max := 100
	i := 0
	lstIntegers := make([]int, 0)
	for i <= 10 {
		lstIntegers = append(lstIntegers, rand.Intn(max-min+1)+min)
		i = i + 1
	}
	fmt.Println("Created Random list of Integers ", lstIntegers)
	fmt.Println("Sorting it by insertion sort")
	fmt.Println("Final Sorted List ", insertionSort(lstIntegers))
}

func insertionSort(lstIntegers []int) []int {

	len := len(lstIntegers)
	for i := 1; i < len; i++ {
		for j := 0; j < i; j++ {
			if lstIntegers[j] > lstIntegers[i] {
				lstIntegers[j], lstIntegers[i] = lstIntegers[i], lstIntegers[j]
			}
		}
	}
	fmt.Println("LstIntegers ",lstIntegers)
	return lstIntegers

}
